catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/TagAggregationVisitor.java

```diff
@@ -0,0 +1,243 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.ui.security.accesscontrol;
+
+import ddf.catalog.data.types.Core;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang.StringUtils;
+import org.opengis.filter.And;
+import org.opengis.filter.Filter;
+import org.opengis.filter.Not;
+import org.opengis.filter.Or;
+import org.opengis.filter.PropertyIsEqualTo;
+import org.opengis.filter.PropertyIsLike;
+import org.opengis.filter.PropertyIsNotEqualTo;
+import org.opengis.filter.expression.Expression;
+import org.opengis.filter.expression.Literal;
+import org.opengis.filter.expression.PropertyName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Visits a filter and aggregates the tags. When the query corresponding to the filter is run, it is
+ * guaranteed any result in the list will have at least one tag found in {@link #getTags()}. If this
+ * cannot be guaranteed, the resultant tag set will be empty.
+ *
+ * <p>Run this visitor against a simplified filter. Running against filters without optimized logic
+ * is not supported. If no other algorithms are available, {@link
+ * org.geotools.filter.visitor.SimplifyingFilterVisitor} can be used.
+ *
+ * <p>Note that {@link PropertyIsNotEqualTo} is not implemented, despite it potentially referencing
+ * metacard tags. That's because aggregation can only effectively operate on what is known to be
+ * true, not necessarily what is known to be false. This is because true predicates narrow down the
+ * result space in deterministic ways without the need of the entire taxonomy or database schema to
+ * compute both sufficient but largely inefficient substitutions.
```

Comment:
✏️ How about rewording for more clarity? 
```
This is because true predicates narrow down the result space in deterministic ways without the need of the entire taxonomy or database schema to compute the truthy alternative for a falsey filter, which would be prohibitively large. 
```

Reply:
Fixed.